esmtools/prediction.py

```diff
@@ -269,485 +276,185 @@ def load_dataset(name, cache=True, data_home=None, **kws):
-# COMPUTE PREDICTABILITY/FORECASTS
-# Highest-level features for computing
-# predictability.
+# COMPARISONS
+# Ways to calculate ensemble spread.
+# Generally from Griffies & Bryan 1997
+# Two different approaches here:
+# - np vectorized from xskillscore (_rmse, _pearson_r) but manually 'stacked'
+#   (_m2m, m2e, ...); supervector is stacked vector of all ensembles and members
+# - xarray vectorized (_mse, _rmse_v, ...) from ensemble variance (_ens_var_against_mean, _..control)
+# Leads to the same results: (metric=_rmse, comparison=c) equals (metric=_rmse_v, comparison=c) for all c in comparisons
 #--------------------------------------------#
-def compute(ds, control, metric=pearson_r, comparison=m2m, anomaly=False, 
-            detrend=False, running=None, varname=None, time_dim='year'):
+
+
+def _m2m(ds, supervector_dim):
     """
-    Add description here.
+    Create two supervectors to compare all members to all other members in turn.
     """
-    supervector_dim = 'svd'
-    if anomaly:
-        _ds = ds - control.mean(time_dim)
-        _control = control - control.mean(time_dim)
-    else:
-        _ds = ds
-        _control = control
-    if detrend:
-        s, i, _, _, _ = et.stats.xr_linregress(_control, time_dim)
-        _control = _control - \
-            (s * _control[time_dim] - _control[time_dim].values[0])
-        _ds = _ds - (s * _ds[time_dim] - _ds[time_dim].values[0])
+    truth_list = []
+    fct_list = []
+    for m in ds.member.values:
+        # drop the member being truth
+        ds_reduced = _drop_members(ds, rmd_member=[m])
+        truth = ds.sel(member=m)
+        for m2 in ds_reduced.member:
+            for e in ds.ensemble:
+                truth_list.append(truth.sel(ensemble=e))
+                fct_list.append(ds_reduced.sel(member=m2, ensemble=e))
+    truth = xr.concat(truth_list, supervector_dim)
+    fct = xr.concat(fct_list, supervector_dim)
+    return fct, truth
 
-    if comparison.__name__ not in ['m2m', 'm2c', 'm2e', 'e2c']:
-        raise ValueError('specify comparison argument')
 
-    if metric.__name__ in ['pearson_r', 'rmse']:
-        fct, truth = comparison(_ds, supervector_dim)
-        res = metric(fct, truth, dim=supervector_dim)
-        res[time_dim]=np.arange(1,res[time_dim].size+1)
-        return res
-    elif metric.__name__ in ['mse', 'rmse_v', 'nrmse', 'nev', 'ppp', 'PPP', 'MSSS', 'uACC']:
-        res = metric(ds, control, comparison, running)
-        res[time_dim]=np.arange(1,res[time_dim].size+1)
-        return res
-    else:
-        raise ValueError('specify metric argument')
+def _ens_var_against_every(ds):
+    """
+    See ens_var_against_mean(ds).
+
+    Only difference is that now distance is evaluated against each ensemble
+    member and then averaged.
+    """
+    var = ds.copy()
+    m = ds.member.size
+    var_list = []
+    for i in range(0, m):
+        var_list.append(
+            ((ds - ds.sel(member=i))**2).sum(dim='member') / (m - 1))
```

Comment:
This should be `ds.isel` since you are doing `for i in range(0, m)`. If the members are listed as `101, 102, 103` or something like that, this won't work.

Reply:
oh, this shows more how ugly and complicated this part of the code is. i'm sorry for you to try to understand this.
I fixed it now, but its still not in a solid way to build more on top of it. ideally, there are the same comparison functions for perfect-model and hindcast setups. 
I will work on this, but now at least it works also for members 100..109.
For now this can go into the code but there will soon be a simplification I hope.

What all those comparisons do is basically broadcasting. I dont understand the concept when I started but thats what those functions basically do. For timeseries broadcast is somehow slower than this concating, but performance only matters on 3d maps, so I will check that out